### DISCUSSION TITLE
v3 Roadmap

### DISCUSSION THREAD

**Discussion by JimBobSquarePants** [SixLabors/ImageSharp#2153] [Ideas]
For v3 I'd like to focus more over streamlining the codebase over implementing new features. 
By doing so we can really focus on making the library as fast as possible while easier to maintain.

Maintenance and performance

- [x] Update License #2151.
- [x] Remove legacy polyfills and compiler conditions.
- [x] Introduce Nullable Reference Types
- [ ] Revisit SIMD pipelines and simplify/optimize (Are we overusing Unsafe API?).
- [ ] Performance pass for RGBA compatible converters.
- [x] Add ARM SIMD support plus CI pipeline.
- [x] Optimize pixel blending with integer arithmetics #1433

New Features

- [x] Decode to target size #2076
- [x] Jpeg encoder complete rewrite #2120
- [ ] Expose convolution API and reimplement [auto matrix linear decomposition](https://github.com/SixLabors/ImageSharp/tree/bdf1b05a12ed6def547df9c3b6c755344ceab58e)
- [x] Investigate OriginalImageFormat #2090
- [ ] Color conversion with ICC profiles #1567
- [x] WrapMemory enhancements #1489

Please note anything else you think I've missed that you feel is important.

**Comment by brianpopow** [comment]
I was thinking about adding encoding for animated webp for v3, but first I am not sure if I will find time soon to do so and second I am not sure if it's that important. Maybe this can wait also for after v3.

> Add ARM SIMD support plus CI pipeline.

While I really would like to have that, I am not 100% convinced that the current PoC in #1994 with QEMU is a very satisfying solution. It will take more the than 1h to run the test just for ARM. Maybe we can find a more clever way to trigger this pipeline only when there are ARM changes or only run ARM specific tests in that pipeline to speed things up. A third option would be to run the pipeline on a custom CI runner with a ARM CPU. Never done that with github, not sure how good that will work, but it for sure will be faster.

**Reply by JimBobSquarePants** [reply]
We can trigger the ARM64 tests when a PR has a certain label.

https://stackoverflow.com/a/62331521

**Reply by brianpopow** [reply]
yeah maybe, that could do it.

**Comment by antonfirsov** [comment]
> Decode to target size

For me this is much more than just #2076 where we are landing an advanced, jpeg-specific feature. What I wish to see is a general `LoadResize()` API with a memory footprint close to the [footprint of vips and magicscaler](https://github.com/saucecontrol/PhotoSauce#magicscaler-efficiency). ImageSharp has to be among the absolute winners.

**Reply by JimBobSquarePants** [reply]
Don't they both use an entirely different pipeline approach though to the other libraries? That would mean significant refactoring. 

Do we have updated numbers for that test anywhere. We must be close to the top 3 now surely?

I agree though that I want the API to be multi format. A jpeg specific API wouldn't match our overall philosophy.

**Reply by antonfirsov** [reply]
> Do we have updated numbers for that test anywhere.

We can produce some with [`LoadResizeSaveParallelMemoryStress`](https://github.com/SixLabors/ImageSharp/blob/main/tests/ImageSharp.Tests.ProfilingSandbox/LoadResizeSaveParallelMemoryStress.cs) but it's not a trivial job since you have to run every library on beefy multi-core desktop or server and analyze graphs manually with WPT.

From what I can recall from my 10 core runs (which was actually stressing parallelism more than Clinton's run with 6 cores):
we went down from 7-10 GB to something 1.3 GB-ish, but vips & magicscaler was around 200 MBs. (But my memory can terribly fail me.)

> Don't they both use an entirely different pipeline approach though to the other libraries

Yes, but we don't have to go fully-pipelined. It's enough if the decoders can somehow utilize the resizer internally, and produce downscaled output as they walk the scanlines.

**Reply by JimBobSquarePants** [reply]
>Yes, but we don't have to go fully-pipelined. It's enough if the decoders can somehow utilize the resizer internally, and produce downscaled output as they walk the scanlines.

I like this. 👍

**Reply by br3aker** [reply]
JFYI we can get really close to libvips in terms of implementation. Jpeg decoding is done via abstract 'convert stride' method, nothing stops us from doing decode-downscale-encode in special converter implementation. That way we won't even need to allocate image itself. But I don't think we actually need to do it this way. IMO `LoadResize` is enough for a general use case, for special use cases people can use netvips.
